proposals/sized-binding-arrays.md

```diff
@@ -0,0 +1,114 @@
+Pipeline layout defaulting is also updated to reflect and set an `arraySize` in the `GPUBindGroupLayouts` it creates.
+
+## WGSL
+
+A new `sized-binding-arrays` language feature is added.
+It can be used for feature detection of both the API-side support and the WGSL-side support.
+
+A new `binding_array<T, N>` type is added that can only be used as a binding and not created directly in shaders.
+
+ - `N` is the size of the array and must be a compile-time constant value or (optionally) an override constant.
+   It is validated to less than or equal to the `arraySize` of the respective `GPUBindGroupLayoutEntry`.
+ - `T` is the type of the binding this array is used for, which is validated against the rest of the contents of the `GPUBindGroupLayoutEntry`.
+
+(TODO for WGSL, can `binding_array` be passed around in function parameters?)
+
+New array access expressions are added for `binding_array`:
+
+ - `e: binding_array<T, N> and i: i32 or u32` -> `e[i]: T`. The result is the value of the `i`th element of the array value `e`.
+    If `i` is outside of the range `[0, N-1]`:
+    
+    - It is a shader-creation error if `i` is a const-expression.
+    - It is a pipeline-creation error if `i` is an override-expression.
+    - Otherwise, any element in the array may be returned.
+
+ - Similar rules for getting a reference to an array element from a memory view to a binding array.
+
+Optionally validation rules for the uniformity, or const-expression-ness of indices into the access expressions are added.
+
+## Extra considerations
+
+WebGPU has composite binding objects with `GPUExternalTexture` that takes multiple slots for limits, and may be represented by multiple bindings in implementations.
+The proposal is to allow `binding_array` for `GPUExternalTexture` as well since it can be translated into per-binding types arrays in shaders.
+
+Texel buffers are a potential future addition to WebGPU and should be considered when deciding which binding types can be arrayed.
+
+## Choices for options outlined above.
+
+To be confirmed:
+
+ - Metal and D3D12 allow non-uniform indexing of bindings.
+ - OpenGL ES allows non-uniform indexing of at least textures.
+ - Vulkan has feature for each individual binding type to allow uniform and non-uniform indexing.
+ - Vulkan uniform indexing of everything but texel buffers is ubiquitous (loses only some LLVMPipe reports in the vulkan gpuinfo query)
+ - Vulkan uniform indexing of texel buffers is added in Vulkan 1.2, but very prevalent there (apart from some Intel devices?)
+ - Vulkan non-uniform indexing of all kinds of bindings is very prevalent but not ubiquitous.
+
+Given that a reasonable workaround exists to support uniform and non-uniform indexing on devices that don't support it, and that the workaround has been used with great success in WebGL, the proposal is to allow non-uniform indexing of all kinds of bindings.
+The workaround looks like the following:
+
+```wgsl
+// Replaces
+var e = myBindingArray[i];
```

Comment:
Probably not a problem yet, but WGSL has a limit of 1023 cases in a switch statement. I see the API currently puts the binding limit at 1000, but also states it is not practical.

Reply:
This wouldn't be a problem as we can expect maxTextureBindingPerStage to stay way below 1000 (bindless wouldn't check against that limit)